mysql-test/main/rowid_filter_innodb.test

```diff
@@ -699,3 +699,61 @@ DROP TABLE t1,t2;
 set global innodb_stats_persistent= @stats.save;
 
 --echo # End of 10.4 tests
+
+--echo #
+--echo # MDEV-31154: Fatal InnoDB error or assertion `!is_v' failure upon multi-update with indexed virtual column
+--echo #
+
+--echo # Test with auto generated Primary Key
+--echo #
+create table t1 (a int, b int as (a/2) virtual, key (b), key (a)) engine=innodb;
+insert into t1(a)
+select (rand(1)*1000)/10 from seq_1_to_1000;
```

Comment:
Sure, I am all for improving the testcase but do we have a way to do it deterministically today ? This has been somewhat hard to repeat and if you see the Elena(filer) also had difficulty providing a repeatable case. If we have a deterministic way it would help everyone. I have elaborated the question more in next comment.

Reply:
The "for update" clause is only to make it follow some path in innodb. It doesn't have much to do with the plan afaiu.

Reply:
> The "for update" clause is only to make it follow some path in innodb

Good, so we've got the path fixed from the innodb side. Maybe `set optimizer_switch='rowid_filter=on';` already fixes the path from the optimizer side? I'd ask the rowid_filter author (Igor Babaev, I suppose) if you have doubts about this.

My only concerns are that the test should be:
1. Deterministic
2. Minimalistic (i.e. fast enough)

Otherwise, if we can't reach both 1 and 2, better not to include it at all, but this will require some explanation then.

Reply:
AFAICT there is no 100% deterministic way in optimizer to enforce rowid_filter. I could be wrong and would be happy to be proven otherwise. `set optimizer_switch='rowid_filter=on';` doesn't guarantee it and the cost calculation  choose different plans if there is no or minimal data in table. The current test is more or less a replica of the another test in the same file introduced by Igor (I have given the details in next comment MDEV-19820) to test the feature for non-virtual columns.

I am fine with any path that we decide to take. Here are the possible options and risks associated with them AFAICS. Either way, I think we should open a testability MDEV in optimizer for this feature.

1. Have a 100% deterministic and minimalistic test. I think it should require some changes in optimizer to  provide that infrastructure. It might cost us to hold the patch till we have a better testing mechanism. We already have another MDEV reported on the same issue by user. https://jira.mariadb.org/browse/MDEV-33558.

2. Remove the test and add some explanation. Attach the test in MDEV. The risk here is not running any test in CI covering the feature and relevant code path. It risks the feature getting broken in future and we come to know only via end user / customer. 
 
3. Have the current test. It covers the feature.
* Not Deterministic: Since it is not 100% deterministic it may not choose the plan in future. In such case the test would fail and we would know that as the plan is also part of the result. Also, similar testcase already exists in the test file and that would possibly fail too and can be handled together.
* Not Minimalistic: It is somewhat hard to define minimalistic test as it is not 100% deterministic. The impact is using more resources in CI.

Reply:
Test improved in 10.4 https://github.com/MariaDB/server/pull/3118